runtime/auth/aws-signing-common/common/test/aws/smithy/kotlin/runtime/auth/awssigning/AwsChunkedByteReadChannelTest.kt

```diff
@@ -0,0 +1,913 @@
+
+        // read the exact chunk length
+        val bytes = awsChunked.readRemaining(
+            CHUNK_SIZE_BYTES + CHUNK_SIZE_BYTES.toString(16).length + 1 +
+                "chunk-signature=".length + 64 + 4 + 1 + 1 + "chunk-signature=".length + 64 + 4 + 2,
+        )
+        val bytesAsString = bytes.decodeToString()
+
+        val chunkSignatures = getChunkSignatures(bytesAsString)
+        assertEquals(chunkSignatures.size, 2) // chunk of data plus an empty chunk
+        var expectedChunkSignature = testSigner.signChunk(data, previousSignature, testSigningConfig).signature
+        assertEquals(expectedChunkSignature.decodeToString(), chunkSignatures[0])
+
+        // the second chunk signature should come from the chunk of zero length
+        expectedChunkSignature = testSigner.signChunk(byteArrayOf(), expectedChunkSignature, testSigningConfig).signature
+        assertEquals(expectedChunkSignature.decodeToString(), chunkSignatures[1])
+
+        val chunkSizes = getChunkSizes(bytesAsString)
+        assertEquals(chunkSizes.size, 2)
+        assertEquals(chunkSizes[0], CHUNK_SIZE_BYTES)
+        assertEquals(chunkSizes[1], 0)
+
+        assertTrue { awsChunked.isClosedForRead } // we've consumed all of the bytes
+    }
+
+    @Test
+    fun testReadRemainingFewerBytes() = runTest {
+        val dataLengthBytes = CHUNK_SIZE_BYTES
+        val data = ByteArray(dataLengthBytes) { 0x7A.toByte() }
+        val chan = SdkByteReadChannel(data)
+        val previousSignature: ByteArray = byteArrayOf()
+        val awsChunked = AwsChunkedByteReadChannel(chan, testSigner, testSigningConfig, previousSignature)
+
+        val numBytesToRead = CHUNK_SIZE_BYTES / 2 // read ~half of the chunk data
+        val bytes = awsChunked.readRemaining(numBytesToRead)
+        val bytesAsString = bytes.decodeToString()
+
+        val chunkSignatures = getChunkSignatures(bytesAsString)
+        assertEquals(chunkSignatures.size, 1) // we should only have the signature of the first chunk
+        val expectedChunkSignature = testSigner.signChunk(data, previousSignature, testSigningConfig).signature
+        assertEquals(expectedChunkSignature.decodeToString(), chunkSignatures[0])
+
+        val chunkSizes = getChunkSizes(bytesAsString)
+        assertEquals(chunkSizes.size, 1)
+        assertEquals(chunkSizes[0], CHUNK_SIZE_BYTES)
+
+        assertFalse { awsChunked.isClosedForRead }
+        assertEquals(bytes.size, numBytesToRead)
+    }
+
+    @Test
+    fun testReadRemainingExcessiveBytes() = runTest {
+        val dataLengthBytes = CHUNK_SIZE_BYTES
+        val data = ByteArray(dataLengthBytes) { 0x7A.toByte() }
+        val chan = SdkByteReadChannel(data)
+        val previousSignature: ByteArray = byteArrayOf()
+        val awsChunked = AwsChunkedByteReadChannel(chan, testSigner, testSigningConfig, previousSignature)
+
+        val numBytesToRead = 131072 // read double of the available chunk data
```

Comment:
since it's a limit why not just make it `Int.MAX_VALUE` or something? The number as written feels oddly specific which doesn't add to understanding the test

Reply:
This is `CHUNK_SIZE_BYTES * 2` -- but I can make that clearer along with the magic numbers in other test cases. `Int.MAX_VALUE` works everywhere except `readRemaining` because we allocate a buffer to hold `limit `bytes and the VM runs out of memory.

```
java.lang.OutOfMemoryError: Requested array size exceeds VM limit
```

I'll change it everywhere except the `testReadRemainingExcessiveBytes` case.

Reply:
Oh I missed that in the code, we really probably shouldn't be allocating the `limit` bytes as we may be way under that in most cases (not to mention it is the [default](https://github.com/awslabs/smithy-kotlin/blob/main/runtime/io/common/src/aws/smithy/kotlin/runtime/io/SdkByteReadChannel.kt#L42)).  

Given we don't ever call `readRemaining()` in any HTTP engines (and never would) and that this will be refactored soon I'm not hugely concerned but I would probably take a second look at this implementation and see if we couldn't do better (e.g. a small optimization would be testing if we are already closed and taking the min(availableForRead, limit)).